model/lists/Service/ListElementsUpdater.php

```diff
@@ -0,0 +1,167 @@
+use core_kernel_classes_Class;
+use tao_helpers_Uri;
+
+class ListElementsUpdater implements ListElementsUpdaterInterface
+{
+    private const HUGE_LIST_MIN_ITEMS = 1000;
+    private const HUGE_LIST_MAX_MEMORY_MB = 500;
+
+    /** @var ValueCollectionService */
+    private $valueCollectionService;
+
+    public function __construct(ValueCollectionService $valueCollectionService)
+    {
+        $this->valueCollectionService = $valueCollectionService;
+    }
+
+    public function setListElements(
+        core_kernel_classes_Class $listClass,
+        array $payload
+    ): bool {
+        $clause = $this->getListSearchInput($listClass);
+        $elementsFromPayload = $this->getElementsFromPayload($payload);
+
+        if ($this->isHugeList($clause, $elementsFromPayload)) {
+            $this->raiseMemoryLimit();
+        }
+
+        $valueCollection = $this->valueCollectionService->findAll($clause);
+
+        foreach ($elementsFromPayload as $key => $value) {
+            $this->addElementToCollection($valueCollection, $key, $value);
+        }
+
+        return $this->valueCollectionService->persist($valueCollection);
+    }
+
+    private function addElementToCollection(
+        ValueCollection $valueCollection,
+        $key,
+        $value
+    ): void {
+        $newUriValue = trim($payload["uri_$key"] ?? '');
+        $element = $this->getValueByUriKey($valueCollection, $key);
+
+        if ($element === null) {
+            $valueCollection->addValue(new Value(null, $newUriValue, $value));
+            return;
+        }
+
+        $element->setLabel($value);
+
+        if ($newUriValue) {
+            $element->setUri($newUriValue);
+        }
+    }
+
+    private function getValueByUriKey(ValueCollection $valueCollection, $key): ?Value
+    {
+        $encodedUri = preg_replace('/^list-element_[0-9]+_/', '', $key);
+        $uri = tao_helpers_Uri::decode($encodedUri);
+
+        if (empty($uri)) {
+            return null;
+        }
+
+        return $valueCollection->extractValueByUri($uri);
+    }
+
+    private function getElementsFromPayload(array $payload): array
+    {
+        return array_filter(
+            $payload,
+            function (string $key): bool {
+                return (bool)preg_match('/^list-element_/', $key);
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+    }
+
+    private function getListSearchInput(
+        core_kernel_classes_Class $listClass
+    ): ValueCollectionSearchInput {
+        return new ValueCollectionSearchInput(
+            (new ValueCollectionSearchRequest())->setValueCollectionUri(
+                $listClass->getUri()
+            )
+        );
+    }
+
+    private function isHugeList(
+        ValueCollectionSearchInput $clause,
+        array $listElements
+    ): bool {
+        $itemsToProcess = max(
+            $this->valueCollectionService->count($clause),
+            count($listElements)
```

Comment:
Will there be a case where `$listElements` are higher than `$this->valueCollectionService->count($clause)`?

Why do you need this check? Btw, lets discuss about this approach on raising memory, I think we are overcomplicating what could be solved with a simple pagination. 

Also, we cannot adjust server memory as we wish to handle this... Lets have a whiteboard session.

Reply:
This is in case `$listElements` may not contain all list items (i.e. if the user is saving elements for a list without loading all its items in the edit form first); in that case `count($listItems)` may be less than the total number of items in the list.

OTOH right now we load all list elements for the list first before persisting (that is needed to match by URI elements provided in the form with existing ones). That means we first load a collection with all existing elements, then we add to that additional elements for those that are new and provided in the payload and persist it,

However, when this particular method `isHugeList()` is called we have not loaded all list items in memory yet.

That means here we'll end up with a list with a size as big as the former one (result of the call to `count()`) OR as big as the former one + new items provided in this request (`RdfValueCollectionRepository` would check for new elements and append them to the list, `RdsValueCollectionRepository` just deletes everything in the DB and then persists new values). Hence we take whichever value is higher (existing size or payload size) as a guess to know if we are likely to need increasing the memory limit.

All of this is used to know if we may have a memory problem because the list which is too big.

Once we've determined that is the case we increase the memory limit, since right now it seems at least one of the repository implementations rely on having all items already in memory.

Reply:
The frontend is now sending just the items that have been changed and we've removed the logic to check and raise the memory limit depending on the request size in the backend, so I'm marking this thread as solved.